File: felicia/examples/learn/topic/common/cc/topic_create_flag.cc
#include "felicia/examples/learn/topic/common/cc/topic_create_flag.h"

#include "third_party/chromium/base/strings/stringprintf.h"

#include "felicia/core/protobuf/channel.pb.h"

namespace felicia {

TopicCreateFlag::TopicCreateFlag() {
  {
    BoolFlag::Builder builder(MakeValueStore(&is_publishing_node_));
    auto flag = builder.SetShortName("-p")
                    .SetHelp(
                        "create publishing node, if set, default: false(create "
                        "subscribing node)")
                    .Build();
    is_publishing_node_flag_ = std::make_unique<BoolFlag>(flag);
  }
  {
    BoolFlag::Builder builder(MakeValueStore(&use_ssl_));
    auto flag =
        builder.SetLongName("--use_ssl")
            .SetHelp("use communication via ssl connection, default: false")
            .Build();
    use_ssl_flag_ = std::make_unique<BoolFlag>(flag);
  }

  {
    StringFlag::Builder builder(MakeValueStore(&topic_));
    auto flag = builder.SetShortName("-t")
                    .SetLongName("--topic")
                    .SetHelp("name for topic, it's required to be set")
                    .Build();
    topic_flag_ = std::make_unique<StringFlag>(flag);
  }
  {
    StringChoicesFlag::Builder builder(MakeValueStore<std::string>(
        &channel_type_, ChannelDef_Type_Name(ChannelDef::CHANNEL_TYPE_TCP),
        Choices<std::string>{
            ChannelDef_Type_Name(ChannelDef::CHANNEL_TYPE_TCP),
            ChannelDef_Type_Name(ChannelDef::CHANNEL_TYPE_UDP),
            ChannelDef_Type_Name(ChannelDef::CHANNEL_TYPE_UDS),
            ChannelDef_Type_Name(ChannelDef::CHANNEL_TYPE_SHM),
        }));
    auto flag =
        builder.SetShortName("-c")
            .SetLongName("--channel_type")
            .SetHelp(base::StringPrintf(
                "protocol to deliver message, it only works for "
                "publishing node (default: %s)",
                ChannelDef_Type_Name(ChannelDef::CHANNEL_TYPE_TCP).c_str()))
            .Build();
    channel_type_flag_ = std::make_unique<StringChoicesFlag>(flag);
  }
}

TopicCreateFlag::~TopicCreateFlag() = default;

bool TopicCreateFlag::Parse(FlagParser& parser) {
  return PARSE_OPTIONAL_FLAG(parser, is_publishing_node_flag_, use_ssl_flag_,
                             name_flag_, topic_flag_, channel_type_flag_);
}

bool TopicCreateFlag::Validate() const {
  return CheckIfFlagWasSet(topic_flag_);
}

}  // namespace felicia

File: felicia/examples/learn/topic/common/python/topic_create_flag.py
from felicia.core.protobuf.channel_pb2 import ChannelDef
from felicia.examples.learn.common.python.node_create_flag import NodeCreateFlag
import felicia_py.command_line_interface as cli


class TopicCreateFlag(NodeCreateFlag):
    def __init__(self):
        super().__init__()
        self.is_publshing_node_flag = cli.BoolFlagBuilder().set_short_name("-p").set_help(
            "create publishing node, if set, default: false(create subscribing node)").build()
        self.use_ssl_flag = cli.BoolFlagBuilder().set_long_name(
            "--use_ssl").set_help("use communication via ssl connection, default: false").build()
        self.topic_flag = cli.StringFlagBuilder().set_short_name("-t").set_long_name(
            "--topic").set_help("name for topic, it's required to be set").build()
        self.channel_type_flag = cli.StringChoicesFlagBuilder(
            ChannelDef.Type.Name(ChannelDef.CHANNEL_TYPE_TCP),
            [
                ChannelDef.Type.Name(ChannelDef.CHANNEL_TYPE_TCP),
                ChannelDef.Type.Name(ChannelDef.CHANNEL_TYPE_UDP),
                ChannelDef.Type.Name(ChannelDef.CHANNEL_TYPE_UDS),
                ChannelDef.Type.Name(ChannelDef.CHANNEL_TYPE_SHM),
            ]).set_short_name(
            "-c").set_long_name("--channel_type").set_help("protocol to deliver message, it only works for publishing node").build()

    def parse(self, flag_parser):
        return self.parse_optional_flags(flag_parser)

    def validate(self):
        return self.check_if_flag_was_set(self.topic_flag)
